File: backend/models/document.py
from models.database import db
from datetime import datetime

class Document(db.Model):
    __tablename__ = 'documents'
    
    id = db.Column(db.Integer, primary_key=True)
    filename = db.Column(db.String(255), nullable=False)
    original_filename = db.Column(db.String(255), nullable=False)
    file_path = db.Column(db.String(500), nullable=False)
    file_type = db.Column(db.String(10), nullable=False)  # pdf, doc, docx, dot, txt
    file_size = db.Column(db.Integer)
    processed = db.Column(db.Boolean, default=False)
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    updated_at = db.Column(db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Relationships
    content = db.relationship('DocumentContent', backref='document', uselist=False)
    bookmarks = db.relationship('Bookmark', backref='document', lazy=True)
    
    def to_dict(self):
        return {
            'id': self.id,
            'filename': self.filename,
            'original_filename': self.original_filename,
            'file_type': self.file_type,
            'file_size': self.file_size,
            'processed': self.processed,
            'created_at': self.created_at.isoformat() if self.created_at else None,
            'updated_at': self.updated_at.isoformat() if self.updated_at else None,
            'content': self.content.to_dict() if self.content else None
        }

class DocumentContent(db.Model):
    __tablename__ = 'document_content'
    
    id = db.Column(db.Integer, primary_key=True)
    document_id = db.Column(db.Integer, db.ForeignKey('documents.id'), nullable=False)
    raw_text = db.Column(db.Text)
    summary = db.Column(db.Text)
    keywords = db.Column(db.Text)  # JSON string of keywords
    word_count = db.Column(db.Integer)
    processing_date = db.Column(db.DateTime, default=datetime.utcnow)
    
    def to_dict(self):
        import json
        return {
            'id': self.id,
            'document_id': self.document_id,
            'summary': self.summary,
            'keywords': json.loads(self.keywords) if self.keywords else [],
            'word_count': self.word_count,
            'processing_date': self.processing_date.isoformat() if self.processing_date else None
        }
</create_file>

File: backend/Models/Document.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace JurisIQ.Backend.Models
{
    public class Document
    {
        [Key]
        public int Id { get; set; }

        [Required]
        [MaxLength(500)]
        public string Title { get; set; } = string.Empty;

        [MaxLength(1000)]
        public string? Description { get; set; }

        [Required]
        [MaxLength(50)]
        public string FileType { get; set; } = string.Empty; // pdf, doc, docx, txt, dot

        [Required]
        [MaxLength(500)]
        public string FilePath { get; set; } = string.Empty;

        [MaxLength(255)]
        public string? CourtName { get; set; }

        [MaxLength(50)]
        public string? CaseNumber { get; set; }

        public DateTime? DecisionDate { get; set; }

        public long FileSize { get; set; }

        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

        public DateTime? UpdatedAt { get; set; }

        public DateTime? ProcessedAt { get; set; }

        public bool IsProcessed { get; set; } = false;

        // Extracted content
        public string? FullText { get; set; }

        public string? Summary { get; set; }

        public string? Keywords { get; set; }

        // Navigation properties
        public virtual ICollection<DocumentKeyword> DocumentKeywords { get; set; } = new List<DocumentKeyword>();
        public virtual ICollection<Bookmark> Bookmarks { get; set; } = new List<Bookmark>();
        public virtual ICollection<DocumentView> Views { get; set; } = new List<DocumentView>();
    }

    public class DocumentKeyword
    {
        [Key]
        public int Id { get; set; }

        [Required]
        [ForeignKey("Document")]
        public int DocumentId { get; set; }
        public virtual Document Document { get; set; } = null!;

        [Required]
        [MaxLength(100)]
        public string Keyword { get; set; } = string.Empty;

        public double RelevanceScore { get; set; } = 0.0;

        public int Frequency { get; set; } = 1;
    }

    public class Bookmark
    {
        [Key]
        public int Id { get; set; }

        [Required]
        [ForeignKey("User")]
        public int UserId { get; set; }
        public virtual User User { get; set; } = null!;

        [Required]
        [ForeignKey("Document")]
        public int DocumentId { get; set; }
        public virtual Document Document { get; set; } = null!;

        [MaxLength(1000)]
        public string? Notes { get; set; }

        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

        // Unique constraint on UserId + DocumentId
    }

    public class DocumentView
    {
        [Key]
        public int Id { get; set; }

        [Required]
        [ForeignKey("Document")]
        public int DocumentId { get; set; }
        public virtual Document Document { get; set; } = null!;

        [Required]
        [ForeignKey("User")]
        public int UserId { get; set; }
        public virtual User User { get; set; } = null!;

        [MaxLength(45)]
        public string? IPAddress { get; set; }

        [MaxLength(255)]
        public string? UserAgent { get; set; }

        public DateTime ViewedAt { get; set; } = DateTime.UtcNow;
    }

    public class SearchHistory
    {
        [Key]
        public int Id { get; set; }

        [Required]
        [ForeignKey("User")]
        public int UserId { get; set; }
        public virtual User User { get; set; } = null!;

        [Required]
        [MaxLength(500)]
        public string Query { get; set; } = string.Empty;

        [MaxLength(100)]
        public string? Filter { get; set; }

        public int ResultCount { get; set; } = 0;

        [MaxLength(45)]
        public string? IPAddress { get; set; }

        public DateTime SearchedAt { get; set; } = DateTime.UtcNow;
    }
}